File: DESC_MATLAB/src/geteqlist_antelope.m
function [eqinfo] = geteqlist_antelope(filename,itype)

if (itype==1) %scec format
	eqlist = getscec(filename);
elseif (itype==2)
	eqlist = gethk(filename);
elseif (itype==3)
    eqlist = getefs(filename);
elseif (itype==4)
    eqlist = getncsn(filename);
elseif (itype==5)
    eqlist = getjma(filename);
elseif (itype==8)
    eqlist = getxyz(filename);
elseif (itype==6) % hypoDD loc file
    eqlist = getdd(filename);
elseif (itype==7) % anss file
    eqlist = getanss(filename);
end

eqinfo(:,1) = eqlist(:,7); %lat
eqinfo(:,2) = eqlist(:,8); %lon
eqinfo(:,3) = eqlist(:,9); %dep
eqinfo(:,4) = eqlist(:,10);%mag
datex = datenum(eqlist(:,1:6));
eqinfo(:,5) = date2epoch(datex);
%% to verify
%	fprintf(1,'%d %d %d %d %d %f\n',eqlist(1,1:6))
%	timenum = epoch2date(eqinfo(1,5))
%	timex = datevec(timenum)
%	fprintf(1,'%d %d %d %d %d %f\n',datevec(datenum(eqlist(1,1:6))))
%for i = 1:neq
%   eqinfo(i,5) = str2epoch(strcat([num2str(eqlist(i,1)),'/',...
%       num2str(eqlist(i,2)),'/',num2str(eqlist(i,3)),'/',...
%       num2str(eqlist(i,4)),':',num2str(eqlist(i,5)),':',...
%       num2str(eqlist(i,6))]));  
%end
eqinfo(:,6) = eqlist(:,11);

function uOutput = getjma(sFilename)
   mData = load(sFilename);
   [M, N ] = size(mData);
   if (N == 10) % then evid exist
       uOutput(:,11) = 1:M;
   elseif (N >= 11)
       uOutput(:,11) = mData(:,11);
   end

   uOutput(:,1:10) = mData(:,1:10);

function uOutput = getxyz(sFilename)
   mData = load(sFilename);
   [M, N ] = size(mData)
   if (N == 10) % then evid exist
       uOutput(:,11) = 1:M;
   elseif (N >= 11)
       uOutput(:,11) = mData(:,11);
   end
   uOutput(:,1:6) = mData(:,1:6);
   uOutput(:,10) = mData(:,7);
   uOutput(:,7:9) = mData(:,8:10);
   %uOutput(:,1:10) = mData(:,1:10);

function uOutput = getscec(sFilename)
    mData = textread(sFilename, '%s', 'delimiter', '\n', 'whitespace', '');
    % Create empty catalog
    uOutput = zeros(length(mData), 10); 
    % Loop through all lines of catalog and convert them
     mData = char(mData);
     l = find( mData == ' ' );
     mData(l) = '0';
     errc = 1;
    try % this is the fast method .... 
        i = 1:length(mData(:,1));   
        [uOutput] = readvaluesscec(mData,i,i);
         
    catch  % this is the line by line method .. 
        for i = 1:length(mData(:,1))
%             if rem(i,100) == 0 ; disp([ num2str(i) ' of ' num2str(length(mData)) ' events processed ']); end          
            try      
                uOutput(i,:) = readvaluesscec(mData,i,1); 
            catch
                errc = errc + 1; 
                if errc == 100
                    if stoploop 
                        return
                    end 
                end      
                disp(['Import: Problem in line ' num2str(i) ' of ' sFilename '. Line ignored.']);
                uOutput(i,:) = uOutput(i,:)*nan;
            end
        end
        
    end
    l = isnan(uOutput(:,1));
    uOutput(l,:) = [];
    nx = find(uOutput(:,7)>100);


function [uOutput] = readvaluesscec(mData,i,k)

% the following is the new SCEC format, but when there is negative depth,
% this will be wrong
uOutput(k,1) = str2num(mData(i,1:4));    % Year
uOutput(k,2) = str2num(mData(i,6:7));    % Month
uOutput(k,3) = str2num(mData(i,9:10));   % Day
uOutput(k,4) = str2num(mData(i,12:13));  % Hour
uOutput(k,5) = str2num(mData(i,15:16));  % Minute
uOutput(k,6) = str2num(mData(i,18:23));  % seconds
uOutput(k,7) = str2num(mData(i,36:43));  % Latitude
uOutput(k,8) = str2num(mData(i,44:51)) ;  % Longitude
uOutput(k,9) = str2num(mData(i,54:58));  % Depth
uOutput(k,10) = str2num(mData(i,27:30));  % Magnitude
uOutput(k,11) = str2num(mData(i,60:67));  % Minute

function uOutput = getefs(sFilename)
    mData = textread(sFilename, '%s', 'delimiter', '\n', 'whitespace', '');
    % Create empty catalog
    uOutput = zeros(length(mData), 10); 
    % Loop through all lines of catalog and convert them
     mData = char(mData);
     l = find( mData == ' ' );
     mData(l) = '0';
     errc = 1;
    try % this is the fast method .... 
        i = 1:length(mData(:,1));   
        [uOutput] = readvaluesefs(mData,i,i);
         
    catch  % this is the line by line method .. 
        for i = 1:length(mData(:,1))
%             if rem(i,100) == 0 ; disp([ num2str(i) ' of ' num2str(length(mData)) ' events processed ']); end          
            try      
                uOutput(i,:) = readvaluesefs(mData,i,1); 
            catch
                errc = errc + 1; 
                if errc == 100
                    if stoploop 
                        return
                    end 
                end      
                disp(['Import: Problem in line ' num2str(i) ' of ' sFilename '. Line ignored.']);
                uOutput(i,:) = uOutput(i,:)*nan;
            end
        end
        
    end
    l = isnan(uOutput(:,1));
    uOutput(l,:) = [];
    nx = find(uOutput(:,7)>100);


function [uOutput] = readvaluesefs(mData,i,k)

% the following is the new SCEC format, but when there is negative depth,
% this will be wrong
uOutput(k,1) = str2num(mData(i,10:13));    % Year
uOutput(k,2) = str2num(mData(i,15:16));    % Month
uOutput(k,3) = str2num(mData(i,18:19));   % Day
uOutput(k,4) = str2num(mData(i,21:22));  % Hour
uOutput(k,5) = str2num(mData(i,24:25));  % Minute
uOutput(k,6) = str2num(mData(i,27:32));  % seconds
uOutput(k,7) = str2num(mData(i,46:53));  % Latitude
uOutput(k,8) = str2num(mData(i,55:64)) ;  % Longitude
uOutput(k,9) = str2num(mData(i,68:72));  % Depth
uOutput(k,10) = str2num(mData(i,37:41));  % Magnitude
uOutput(k,11) = str2num(mData(i,1:8));  % scecid


function uOutput = gethk(sFilename)
    mData = textread(sFilename, '%s', 'delimiter', '\n', 'whitespace', '');
    % Create empty catalog
    uOutput = zeros(length(mData),11);
    Error_lines = []; 
    % Loop through all lines of catalog and convert them
    mData = char(mData);
    l = find( mData == ' ' );
    mData(l) = '0';
    errc = 1;
    try % this is the fast method .... 
        i = 1:length(mData(:,1));   
        [uOutput] = readvalueshk(mData,i,i);
         
    catch  % this is the line by line method .. 
        for i = 1:length(mData(:,1))
            if rem(i,100) == 0 ; disp([ num2str(i) ' of ' num2str(length(mData)) ' events processed ']); end          
            try      
                uOutput(i,:) = readvalueshk(mData,i,1); 
            catch
                errc = errc + 1; 
                if errc == 100
                    if stoploop 
                        return
                    end 
                end      
                disp(['Import: Problem in line ' num2str(i) ' of ' sFilename '. Line ignored.']);
                uOutput(i,:) = uOutput(i,:)*nan;
                Error_lines = [Error_lines; i];
            end
        end
        
    end
    l = isnan(uOutput(:,1));
    uOutput(l,:) = [];
    
function [uOutput] = readvalueshk(mData,i,k)

uOutput(k,1) = str2num(mData(i,(1:4)));    % Year
uOutput(k,2) = str2num(mData(i,(6:7)));    % Month
uOutput(k,3) = str2num(mData(i,(9:10)));   % Day
uOutput(k,4) = str2num(mData(i,(12:13)));  % Hour
uOutput(k,5) = str2num(mData(i,(15:16)));  % Minute
uOutput(k,6) = str2num(mData(i,(18:23)));  % Second
uOutput(k,7) = str2num(mData(i,(35:42)));  % Latitude [deg]
uOutput(k,8) = str2num(mData(i,(44:53)));  % Longitude [deg]
uOutput(k,9) = str2num(mData(i,(56:60)));  % Depth [km]
uOutput(k,10) = str2num(mData(i,(63:68)));  % Magnitude
uOutput(k,11) = str2num(mData(i,(25:33))); % SCSN cuspid (up to 9 digits)


function uOutput = getncsn(sFilename)
    mData = textread(sFilename, '%s', 'delimiter', '\n', 'whitespace', '');
    % Create empty catalog
    uOutput = zeros(length(mData), 10); 
    % Loop through all lines of catalog and convert them
     mData = char(mData);
     l = find( mData == ' ' );
     mData(l) = '0';
     errc = 1;
    try % this is the fast method .... 
        i = 1:length(mData(:,1));   
        [uOutput] = readvaluesncsn(mData,i,i);
         
    catch  % this is the line by line method .. 
        for i = 1:length(mData(:,1))
%             if rem(i,100) == 0 ; disp([ num2str(i) ' of ' num2str(length(mData)) ' events processed ']); end          
            try      
                uOutput(i,:) = readvaluesncsn(mData,i,1); 
            catch
                errc = errc + 1; 
                if errc == 100
                    if stoploop 
                        return
                    end 
                end      
                disp(['Import: Problem in line ' num2str(i) ' of ' sFilename '. Line ignored.']);
                uOutput(i,:) = uOutput(i,:)*nan;
            end
        end
        
    end
    l = isnan(uOutput(:,1));
    uOutput(l,:) = [];
    nx = find(uOutput(:,7)>100);


function [uOutput] = readvaluesncsn(mData,i,k)

% the following is the new SCEC format, but when there is negative depth,
% this will be wrong
uOutput(k,1) = str2num(mData(i,1:4));    % Year
uOutput(k,2) = str2num(mData(i,6:7));    % Month
uOutput(k,3) = str2num(mData(i,9:10));   % Day
uOutput(k,4) = str2num(mData(i,12:13));  % Hour
uOutput(k,5) = str2num(mData(i,15:16));  % Minute
uOutput(k,6) = str2num(mData(i,18:22));  % seconds
uOutput(k,7) = str2num(mData(i,25:33));  % Latitude
uOutput(k,8) = str2num(mData(i,34:44)) ;  % Longitude
uOutput(k,9) = str2num(mData(i,46:52));  % Depth
uOutput(k,10) = str2num(mData(i,54:58));  % Magnitude
uOutput(k,11) = str2num(mData(i,90:97));  % ncsn id

function uOutput = getdd(sFilename)
    mData = load(sFilename);
    uOutput(:,1) = mData(:,11); %year
    uOutput(:,2) = mData(:,12);
    uOutput(:,3) = mData(:,13);
    uOutput(:,4) = mData(:,14);
    uOutput(:,5) = mData(:,15);
    uOutput(:,6) = mData(:,16); %second
    uOutput(:,7) = mData(:,2); %lat
    uOutput(:,8) = mData(:,3);
    uOutput(:,9) = mData(:,4); %dep
    uOutput(:,10) = mData(:,17); %mag
    uOutput(:,11) = mData(:,1); %id

    
function uOutput = getanss(sFilename)
mData = textread(sFilename, '%s', 'delimiter', '\n', 'whitespace', '');
    % Create empty catalog
    uOutput = zeros(length(mData), 11); 
    % Loop through all lines of catalog and convert them
     mData = char(mData);
     l = find( mData == ' ' );
     mData(l) = '0';
     errc = 1;
    try % this is the fast method .... 
        i = 1:length(mData(:,1));   
        [uOutput] = readvaluesanss(mData,i,i);
         
    catch  % this is the line by line method .. 
        for i = 1:length(mData(:,1))
%             if rem(i,100) == 0 ; disp([ num2str(i) ' of ' num2str(length(mData)) ' events processed ']); end          
            try      
                uOutput(i,:) = readvaluesanss(mData,i,1); 
            catch
                errc = errc + 1; 
                if errc == 100 
                    if stoploop 
                        return
                    end 
                end      
                disp(['Import: Problem in line ' num2str(i) ' of ' sFilename '. Line ignored.']);
                uOutput(i,:) = uOutput(i,:)*nan;
            end
        end
        
    end
    l = isnan(uOutput(:,1));
    uOutput(l,:) = [];
    nx = find(uOutput(:,7)>100);


function [uOutput] = readvaluesanss(mData,i,k)

% the following is the new SCEC format, but when there is negative depth,
% this will be wrong
uOutput(k,1) = str2num(mData(i,1:4));    % Year
uOutput(k,2) = str2num(mData(i,6:7));    % Month
uOutput(k,3) = str2num(mData(i,9:10));   % Day
uOutput(k,4) = str2num(mData(i,12:13));  % Hour
uOutput(k,5) = str2num(mData(i,15:16));  % Minute
uOutput(k,6) = str2num(mData(i,18:22));  % seconds
uOutput(k,7) = str2num(mData(i,23:31));  % Latitude
uOutput(k,8) = str2num(mData(i,32:41)) ;  % Longitude
uOutput(k,9) = str2num(mData(i,42:48));  % Depth
uOutput(k,10) = str2num(mData(i,49:54));  % Magnitude
uOutput(k,11) = str2num(mData(i,84:96));  % ncsn id

%%%%%%%%%%%%%%%%%%%%%%%%%%%%% 
%%%%    DO NOT CHANGE %%%%%%%%%%%

function [mystop] = stoploop()

ButtonName=questdlg('More than 100 lines could not be read. Continue?', ...
    'Interrupt?', ...
    'Yes','No','Nope');

switch ButtonName
case 'Yes'
    disp('going on');
    mystop = 0;
case 'No'
    mystop = 1;
end % switch

File: DESC_Python/src/geteqlist_antelope.py
import numpy as np


def _loadtxt_2d(filename):
    """Load numeric text and ensure a 2-D shape even for single-line files."""
    mdata = np.loadtxt(filename)
    if mdata.ndim == 1:
        mdata = mdata.reshape(1, -1)
    return mdata


def geteqlist_antelope(filename, itype):
    """
    Read earthquake catalog in various formats.
    
    Parameters
    ----------
    filename : str
        Path to catalog file.
    itype : int
        Format type:
        1 = SCEC, 2 = HK, 3 = EFS, 4 = NCSN, 5 = JMA, 
        6 = hypoDD, 7 = ANSS, 8 = XYZ
    
    Returns
    -------
    eqlist : ndarray
        Array with columns: [year, month, day, hour, minute, second, 
                             lat, lon, depth, mag, evid]
    """
    if itype == 1:
        eqlist = _get_scec(filename)
    elif itype == 2:
        eqlist = _get_hk(filename)
    elif itype == 3:
        eqlist = _get_efs(filename)
    elif itype == 4:
        eqlist = _get_ncsn(filename)
    elif itype == 5:
        eqlist = _get_jma(filename)
    elif itype == 6:
        eqlist = _get_dd(filename)
    elif itype == 7:
        eqlist = _get_anss(filename)
    elif itype == 8:
        eqlist = _get_xyz(filename)
    else:
        raise ValueError(f"Unknown catalog type: {itype}")
    
    return eqlist


def _get_jma(filename):
    """JMA format: load numeric data from file."""
    mData = _loadtxt_2d(filename)
    M, N = mData.shape
    uOutput = np.zeros((M, 11))
    
    if N == 10:
        uOutput[:, 10] = np.arange(1, M + 1)
    elif N >= 11:
        uOutput[:, 10] = mData[:, 10]
    
    uOutput[:, :10] = mData[:, :10]
    return uOutput


def _get_xyz(filename):
    """XYZ format: columns rearranged [yr mon day hr min sec, mag, lat lon dep, evid]."""
    mData = _loadtxt_2d(filename)
    M, N = mData.shape
    uOutput = np.zeros((M, 11))
    
    if N == 10:
        uOutput[:, 10] = np.arange(1, M + 1)
    elif N >= 11:
        uOutput[:, 10] = mData[:, 10]
    
    uOutput[:, :6] = mData[:, :6]
    uOutput[:, 9] = mData[:, 6]      # mag
    uOutput[:, 6:9] = mData[:, 7:10] # lat, lon, dep
    return uOutput


def _get_scec(filename):
    """SCEC format: fixed-width text parser."""
    uOutput = []
    
    with open(filename, 'r') as f:
        for line_no, line in enumerate(f, 1):
            line = line.replace(' ', '0')
            try:
                row = _read_values_scec(line)
                if not np.isnan(row[0]):
                    uOutput.append(row)
            except Exception as e:
                print(f"Import: Problem in line {line_no} of {filename}. Line ignored.")
                continue
    
    uOutput = np.array(uOutput)
    # Filter invalid latitudes
    valid = uOutput[:, 6] <= 100
    return uOutput[valid]


def _read_values_scec(line):
    """Parse SCEC fixed-width format."""
    row = np.zeros(11)
    row[0] = float(line[0:4])       # year
    row[1] = float(line[5:7])       # month
    row[2] = float(line[8:10])      # day
    row[3] = float(line[11:13])     # hour
    row[4] = float(line[14:16])     # minute
    row[5] = float(line[17:23])     # second
    row[6] = float(line[35:43])     # latitude
    row[7] = float(line[43:51])     # longitude
    row[8] = float(line[53:58])     # depth
    row[9] = float(line[26:30])     # magnitude
    row[10] = float(line[59:67])    # event ID
    return row


def _get_efs(filename):
    """EFS format: fixed-width text parser."""
    uOutput = []
    
    with open(filename, 'r') as f:
        for line_no, line in enumerate(f, 1):
            line = line.replace(' ', '0')
            try:
                row = _read_values_efs(line)
                if not np.isnan(row[0]):
                    uOutput.append(row)
            except Exception as e:
                print(f"Import: Problem in line {line_no} of {filename}. Line ignored.")
                continue
    
    uOutput = np.array(uOutput)
    valid = uOutput[:, 6] <= 100
    return uOutput[valid]


def _read_values_efs(line):
    """Parse EFS fixed-width format."""
    row = np.zeros(11)
    row[0] = float(line[9:13])      # year
    row[1] = float(line[14:16])     # month
    row[2] = float(line[17:19])     # day
    row[3] = float(line[20:22])     # hour
    row[4] = float(line[23:25])     # minute
    row[5] = float(line[26:32])     # second
    row[6] = float(line[45:53])     # latitude
    row[7] = float(line[54:64])     # longitude
    row[8] = float(line[67:72])     # depth
    row[9] = float(line[36:41])     # magnitude
    row[10] = float(line[0:8])      # scecid
    return row


def _get_hk(filename):
    """HK format: fixed-width text parser."""
    uOutput = []
    
    with open(filename, 'r') as f:
        for line_no, line in enumerate(f, 1):
            line = line.replace(' ', '0')
            try:
                row = _read_values_hk(line)
                if not np.isnan(row[0]):
                    uOutput.append(row)
            except Exception as e:
                print(f"Import: Problem in line {line_no} of {filename}. Line ignored.")
                continue
    
    return np.array(uOutput)


def _read_values_hk(line):
    """Parse HK fixed-width format."""
    row = np.zeros(11)
    row[0] = float(line[0:4])       # year
    row[1] = float(line[5:7])       # month
    row[2] = float(line[8:10])      # day
    row[3] = float(line[11:13])     # hour
    row[4] = float(line[14:16])     # minute
    row[5] = float(line[17:23])     # second
    row[6] = float(line[34:42])     # latitude
    row[7] = float(line[43:53])     # longitude
    row[8] = float(line[55:60])     # depth
    row[9] = float(line[62:68])     # magnitude
    row[10] = float(line[24:33])    # cuspid
    return row


def _get_ncsn(filename):
    """NCSN format: fixed-width text parser."""
    uOutput = []
    
    with open(filename, 'r') as f:
        for line_no, line in enumerate(f, 1):
            line = line.replace(' ', '0')
            try:
                row = _read_values_ncsn(line)
                if not np.isnan(row[0]):
                    uOutput.append(row)
            except Exception as e:
                print(f"Import: Problem in line {line_no} of {filename}. Line ignored.")
                continue
    
    uOutput = np.array(uOutput)
    valid = uOutput[:, 6] <= 100
    return uOutput[valid]


def _read_values_ncsn(line):
    """Parse NCSN fixed-width format."""
    row = np.zeros(11)
    row[0] = float(line[0:4])       # year
    row[1] = float(line[5:7])       # month
    row[2] = float(line[8:10])      # day
    row[3] = float(line[11:13])     # hour
    row[4] = float(line[14:16])     # minute
    row[5] = float(line[17:22])     # second
    row[6] = float(line[24:33])     # latitude
    row[7] = float(line[33:44])     # longitude
    row[8] = float(line[45:52])     # depth
    row[9] = float(line[53:58])     # magnitude
    row[10] = float(line[89:97])    # ncsn id
    return row


def _get_dd(filename):
    """hypoDD format: load and rearrange columns."""
    mData = _loadtxt_2d(filename)
    uOutput = np.zeros((mData.shape[0], 11))
    uOutput[:, 0] = mData[:, 10]    # year
    uOutput[:, 1] = mData[:, 11]    # month
    uOutput[:, 2] = mData[:, 12]    # day
    uOutput[:, 3] = mData[:, 13]    # hour
    uOutput[:, 4] = mData[:, 14]    # minute
    uOutput[:, 5] = mData[:, 15]    # second
    uOutput[:, 6] = mData[:, 1]     # lat
    uOutput[:, 7] = mData[:, 2]     # lon
    uOutput[:, 8] = mData[:, 3]     # depth
    uOutput[:, 9] = mData[:, 16]    # mag
    uOutput[:, 10] = mData[:, 0]    # id
    return uOutput


def _get_anss(filename):
    """ANSS format: fixed-width text parser."""
    uOutput = []
    
    with open(filename, 'r') as f:
        for line_no, line in enumerate(f, 1):
            line = line.replace(' ', '0')
            try:
                row = _read_values_anss(line)
                if not np.isnan(row[0]):
                    uOutput.append(row)
            except Exception as e:
                print(f"Import: Problem in line {line_no} of {filename}. Line ignored.")
                continue
    
    uOutput = np.array(uOutput)
    valid = uOutput[:, 6] <= 100
    return uOutput[valid]


def _read_values_anss(line):
    """Parse ANSS fixed-width format."""
    row = np.zeros(11)
    row[0] = float(line[0:4])       # year
    row[1] = float(line[5:7])       # month
    row[2] = float(line[8:10])      # day
    row[3] = float(line[11:13])     # hour
    row[4] = float(line[14:16])     # minute
    row[5] = float(line[17:22])     # second
    row[6] = float(line[22:31])     # latitude
    row[7] = float(line[31:41])     # longitude
    row[8] = float(line[41:48])     # depth
    row[9] = float(line[48:54])     # magnitude
    row[10] = float(line[83:96])    # ncsn id
    return row
